src/lib/key-provider.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright (c) 2017, [Ribose Inc](https://www.ribose.com).
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1.  Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ * 2.  Redistributions in binary form must reproduce the above copyright notice,
+ *     this list of conditions and the following disclaimer in the documentation
+ *     and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef RNP_KEY_PROVIDER_H
+#define RNP_KEY_PROVIDER_H
+
+#include <rnp/rnp_types.h>
+#include <rnp/rnp_sdk.h>
+#include "pass-provider.h"
+
+typedef struct pgp_key_t pgp_key_t;
+
+typedef enum {
+    PGP_KEY_SEARCH_KEYID,
+    PGP_KEY_SEARCH_GRIP,
+    PGP_KEY_SEARCH_USERID
+} pgp_key_search_t;
+
+typedef struct pgp_key_request_ctx_t {
+    pgp_passphrase_provider_t *pass_provider;
```

Comment:
I think it's better to remove this field and use the passphrase provider within `rnp_t`. This structure should contain the minimal amount of info for retrieving a key.

Reply:
I added rnp_t parameter only because we do not have good separated keyring object. 
Currently streamed methods do not make any use of rnp_t, and the only information they need is password provider and key provider. So imho it is better to do it in this way, later on separating and abstracting keyring object out of the rnp_t. What do you think?

Reply:
Hm, got to another point. Callee needs to distinguish whether key is unavailable or password is incorrect. So I think we should remove unlocking from this code at all - it will be done by the callee.